python/src/trezorlib/cardano.py

```diff
@@ -212,21 +257,58 @@ def _parse_tokens(tokens) -> List[messages.CardanoToken]:
 def _parse_address_parameters(
     address_parameters,
 ) -> messages.CardanoAddressParametersType:
-    if "path" not in address_parameters:
+    if "addressType" not in address_parameters:
         raise ValueError(INCOMPLETE_OUTPUT_ERROR_MESSAGE)
 
-    staking_key_hash_bytes = None
-    if "stakingKeyHash" in address_parameters:
-        staking_key_hash_bytes = bytes.fromhex(address_parameters.get("stakingKeyHash"))
+    path = tools.parse_path(address_parameters.get("path"))
+    staking_path = tools.parse_path(address_parameters.get("stakingPath"))
+    staking_key_hash_bytes = parse_optional_bytes(
+        address_parameters.get("stakingKeyHash")
+    )
+    script_payment_hash = parse_optional_bytes(
+        address_parameters.get("scriptPaymentHash")
+    )
+    script_staking_hash = parse_optional_bytes(
+        address_parameters.get("scriptStakingHash")
+    )
 
     return create_address_parameters(
         int(address_parameters["addressType"]),
-        tools.parse_path(address_parameters["path"]),
-        tools.parse_path(address_parameters.get("stakingPath")),
+        path,
+        staking_path,
         staking_key_hash_bytes,
         address_parameters.get("blockIndex"),
         address_parameters.get("txIndex"),
         address_parameters.get("certificateIndex"),
+        script_payment_hash,
+        script_staking_hash,
+    )
+
+
+def parse_native_script(native_script) -> messages.CardanoNativeScript:
+    if "type" not in native_script:
+        raise ValueError("Script is missing some fields")
+
+    type = native_script["type"]
+    scripts = [
+        parse_native_script(sub_script)
+        for sub_script in native_script.get("scripts", ())
+    ]
+
+    key_hash = parse_optional_bytes(native_script.get("key_hash"))
+    key_path = tools.parse_path(native_script.get("key_path"))
+    required_signatures_count = native_script.get("required_signatures_count")
+    invalid_before = native_script.get("invalid_before")
```

Comment:
We are not using `int()` while we do e.g. in `_parse_tokens`.

Reply:
https://github.com/vacuumlabs/trezor-firmware/pull/66/commits/8293956d38bf2012776f6aee0838e62f3693dcce#diff-aa355fcce844deffafbd696c1d414eb1c0198b0af2726710c7b8d8f7fda577cdR304 `parse_optional_int` should perhaps be also used in other places, but I'll leave that out of this PR. The whole file should perhaps be refactored at some point.